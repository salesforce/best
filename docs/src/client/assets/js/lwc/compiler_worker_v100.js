/* eslint-disable no-undef */
// -- globalThis polyfill -----------------------------------------------------
(function() {
    if (typeof globalThis === 'object') return;
    // eslint-disable-next-line no-extend-native
    Object.defineProperty(Object.prototype, '__magic__', {
        get: function() {
            return this;
        },
        configurable: true,
    });
    __magic__.globalThis = __magic__;
    delete Object.prototype.__magic__;
})();

// -- https://github.com/nolanlawson/promise-worker/blob/master/register.js
function isPromise(obj) {
    // via https://unpkg.com/is-promise@2.1.0/index.js
    return (
        !!obj &&
        (typeof obj === 'object' || typeof obj === 'function') &&
        typeof obj.then === 'function'
    );
}

function registerPromiseWorker(callback) {
    function postOutgoingMessage(e, messageId, error, result) {
        function postMessage(msg) {
            /* istanbul ignore if */
            if (typeof self.postMessage !== 'function') {
                // service worker
                e.ports[0].postMessage(msg);
            } else {
                // web worker
                self.postMessage(msg);
            }
        }
        if (error) {
            /* istanbul ignore else */
            if (typeof console !== 'undefined' && 'error' in console) {
                // This is to make errors easier to debug. I think it's important
                // enough to just leave here without giving the user an option
                // to silence it.
                // eslint-disable-next-line no-console
                console.error('Worker caught an error:', error);
            }
            postMessage([
                messageId,
                {
                    message: error.message,
                },
            ]);
        } else {
            postMessage([messageId, null, result]);
        }
    }

    function tryCatchFunc(callback, message) {
        try {
            return { res: callback(message) };
        } catch (e) {
            return { err: e };
        }
    }

    function handleIncomingMessage(e, callback, messageId, message) {
        var result = tryCatchFunc(callback, message);

        if (result.err) {
            postOutgoingMessage(e, messageId, result.err);
        } else if (!isPromise(result.res)) {
            postOutgoingMessage(e, messageId, null, result.res);
        } else {
            result.res.then(
                function(finalResult) {
                    postOutgoingMessage(e, messageId, null, finalResult);
                },
                function(finalError) {
                    postOutgoingMessage(e, messageId, finalError);
                },
            );
        }
    }

    function onIncomingMessage(e) {
        var payload = e.data;
        if (!Array.isArray(payload) || payload.length !== 2) {
            // message doens't match communication format; ignore
            return;
        }
        var messageId = payload[0];
        var message = payload[1];

        if (typeof callback !== 'function') {
            postOutgoingMessage(e, messageId, new Error('Please pass a function into register().'));
        } else {
            handleIncomingMessage(e, callback, messageId, message);
        }
    }

    self.addEventListener('message', onIncomingMessage);
}

// -- main worker code -----------------------------------------------------

importScripts('/assets/js/lwc/compiler_v100.js');
const compile = globalThis.compile;

registerPromiseWorker(async function({ operation, options }) {
    if (operation === 'ping') {
        return 'ready';
    }

    if (operation === 'compile') {
        const { input, virtualFileSystem } = options;
        const result = await compile(input, virtualFileSystem);
        return result.output[0].code;
    }
});
