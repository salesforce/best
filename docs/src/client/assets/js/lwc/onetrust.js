/*globals console */
/* eslint-disable */
var SfdcWwwBase = SfdcWwwBase || {};

var oneTrustComponent = (function() {
    'use strict';

    // oneTrustContainer gets set in waitForOneTrustContainer()
    var oneTrustContainer,
        count = 0;
    var Keyboard = {
        TAB: 9,
        ESCAPE: 27,
        ENTER: 13,
        SPACE: 32,
        LEFT: 37,
        RIGHT: 39,
    };

    function changeButtons() {
        var categoryItems =
            oneTrustContainer && oneTrustContainer.querySelectorAll('.category-menu-switch-handler')
                ? oneTrustContainer.querySelectorAll('.category-menu-switch-handler')
                : false;
        var saveCookiesButton =
            oneTrustContainer && oneTrustContainer.querySelector('.save-preference-btn-handler')
                ? oneTrustContainer.querySelector('.save-preference-btn-handler')
                : false;
        var acceptCookiesButton =
            oneTrustContainer && oneTrustContainer.querySelector('#accept-recommended-btn-handler')
                ? oneTrustContainer.querySelector('#accept-recommended-btn-handler')
                : false;
        var generalContentArea =
            oneTrustContainer && oneTrustContainer.querySelector('#ot-tab-desc')
                ? oneTrustContainer.querySelector('#ot-tab-desc')
                : false;

        // Update backbutton in cookie list to move span inside of button
        var listSection =
            oneTrustContainer && oneTrustContainer.querySelector('#ot-pc-lst')
                ? oneTrustContainer.querySelector('#ot-pc-lst')
                : false;
        var headingSection =
            listSection && listSection.querySelector('#ot-lst-title')
                ? listSection.querySelector('#ot-lst-title')
                : false;
        var backButton =
            headingSection && headingSection.querySelector('.back-btn-handler')
                ? headingSection.querySelector('.back-btn-handler')
                : false;
        var headingSecSpan =
            headingSection && headingSection.getElementsByTagName('span')
                ? headingSection.getElementsByTagName('span')
                : false;

        // Move span inside of button so all can be selected or clicked, not just tiny arrow
        if (backButton && headingSecSpan) backButton.appendChild(headingSecSpan[0]);

        // Setup button toggles to show/hide save all button
        function toggleButtons(categoryItem) {
            if (saveCookiesButton && acceptCookiesButton) {
                if (typeof categoryItem.parentElement.dataset.optanongroupid === 'undefined') {
                    if (saveCookiesButton.classList.contains('visible'))
                        saveCookiesButton.classList.remove('visible');
                    if (acceptCookiesButton.classList.contains('optanon-ghost-button'))
                        acceptCookiesButton.classList.remove('optanon-ghost-button');
                } else {
                    saveCookiesButton.classList.add('visible');
                    acceptCookiesButton.classList.add('optanon-ghost-button');
                }
            }
        }

        // Loop over links to add click and keydown events to call toggleButtons()
        for (let i = 0; i < categoryItems.length; i++) {
            // Click events
            categoryItems[i].addEventListener('click', function(e) {
                toggleButtons(e.currentTarget);
            });

            // Add keyboard navigation event to toggle buttons
            categoryItems[i].addEventListener('keydown', function(e) {
                // Arrow keys - left and right
                if (e.keyCode === Keyboard.LEFT || e.keyCode === Keyboard.RIGHT)
                    toggleArrowKeyDirection();
            });
        }

        // OneTrust tool not fully accessible. Needs a little help with tab key and shift + tab keys
        if (
            generalContentArea &&
            saveCookiesButton &&
            acceptCookiesButton &&
            categoryItems.length
        ) {
            for (let i = 0; i < categoryItems.length; i++) {
                if (i === 0) {
                    // Add listener to first item in list only.
                    categoryItems[0].addEventListener('keydown', function(e) {
                        if (e.shiftKey === true && e.keyCode === Keyboard.TAB) {
                            // Look for shift + tab key
                            // Time out needed to prevent collision with OneTrust listener on tab key
                            setTimeout(function() {
                                acceptCookiesButton.focus();
                            }, 100);
                        }
                        if (e.shiftKey === false && e.keyCode === Keyboard.TAB) {
                            // Look for tab key
                            // Time out needed to prevent collision with OneTrust listener on tab key
                            setTimeout(function() {
                                generalContentArea.focus();
                            }, 100);
                        }
                    });
                } else {
                    // Captures focus to only OneTrust tool. Prevents shift tab from going into page level elements
                    categoryItems[i].addEventListener('keydown', function(e) {
                        if (e.shiftKey === true && e.keyCode === Keyboard.TAB) {
                            // Look for shift + tab key
                            // Time out needed to prevent collision with OneTrust listener on tab key
                            setTimeout(function() {
                                acceptCookiesButton.focus();
                            }, 100);
                        }
                    });
                }
            }
        }

        // OneTrust activates menu items with left and right arrow keys.
        // We just need to look for the active item and pass it into toggleButtons()
        function toggleArrowKeyDirection() {
            // Check for active menu then pass it into toggle button function
            for (let i = 0; i < categoryItems.length; i++) {
                if (categoryItems[i].classList.contains('ot-active-menu')) {
                    toggleButtons(categoryItems[i]);
                }
            }
        }
    }

    function waitForOneTrustContainer() {
        // Container may not be on the page at the time of page load, wait for it
        var oneTrustStatusInterval = setInterval(function() {
            // Try to set one trust container
            oneTrustContainer = document.querySelector('#onetrust-pc-sdk');
            // Check that we have a container
            if (
                oneTrustContainer !== undefined &&
                oneTrustContainer !== null &&
                oneTrustContainer
            ) {
                // If we have a container, proceed with change button wireup. Then clear interval.
                changeButtons();
                clearInterval(oneTrustStatusInterval);
            }

            if (count++ > 10) {
                // timeout if this takes more than 5 sec
                clearInterval(oneTrustStatusInterval);
            }
        }, 500);
    }

    function init() {
        waitForOneTrustContainer();
    }

    return {
        init: init,
        oneTrustComponent: oneTrustComponent,
    };
})();

function runOneTrustComponent() {
    oneTrustComponent.init();
}

// In case the document is already rendered
if (document.readyState != 'loading') runOneTrustComponent();
// Modern browsers
else if (document.addEventListener)
    document.addEventListener('DOMContentLoaded', runOneTrustComponent);
// IE <= 8
else
    document.attachEvent('onreadystatechange', function() {
        if (document.readyState == 'complete') runOneTrustComponent();
    });