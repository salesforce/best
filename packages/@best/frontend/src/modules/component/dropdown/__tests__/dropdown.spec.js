import { createElement } from 'lwc';
import Dropdown from 'component/dropdown';

describe('component-dropdown', () => {
    afterEach(() => {
        // The jsdom instance is shared across test cases in a single file so reset the DOM
        while (document.body.firstChild) {
            document.body.removeChild(document.body.firstChild);
        }
    });

    it('display correct text and data-index with one selected item', () => {
        const selectedItem = {
            title: 'First Selected Item',
            id: 'item-one'
        }

        const items = [
            selectedItem,
            {
                title: 'Second Item',
                id: 'item-two'
            }, {
                title: 'Third Item',
                id: 'item-three'
            }
        ]

        const element = createElement('component-dropdown', { is: Dropdown })
        element.options = {
            multiple: false,
            items,
            selectedItems: [
                selectedItem
            ]
        }

        document.body.appendChild(element)

        return Promise.resolve().then(() => {
            expect(element).toMatchSnapshot();
            const selectedItemElement = element.shadowRoot.querySelector('.selected-item');
            expect(selectedItemElement.textContent).toBe(selectedItem.title);
            expect(parseInt(selectedItemElement.dataset.index, 10)).toBe(0);
        })
    })

    it('displays no selected items when none are provided', () => {
        const items = [
            {
                title: 'First Selected Item',
                id: 'item-one'
            },
            {
                title: 'Second Item',
                id: 'item-two'
            }, {
                title: 'Third Item',
                id: 'item-three'
            }
        ]

        const element = createElement('component-dropdown', { is: Dropdown })
        element.options = {
            multiple: false,
            items,
            selectedItems: []
        }

        document.body.appendChild(element)

        return Promise.resolve().then(() => {
            expect(element).toMatchSnapshot();
            const allSelectedItemElements = element.shadowRoot.querySelectorAll('.selected-item');
            expect(allSelectedItemElements).toHaveLength(0);
        })
    })

    it('displays no items when none are provided', () => {
        const element = createElement('component-dropdown', { is: Dropdown })
        element.options = {
            multiple: false,
            items: [],
            selectedItems: []
        }

        document.body.appendChild(element)

        return Promise.resolve().then(() => {
            expect(element).toMatchSnapshot();
            const allItemElements = element.shadowRoot.querySelectorAll('.item');
            expect(allItemElements).toHaveLength(0);
        })
    })

    it('displays items in correct order', () => {
        const items = [
            {
                title: 'First Selected Item',
                id: 'item-one'
            },
            {
                title: 'Second Item',
                id: 'item-two'
            }, {
                title: 'Third Item',
                id: 'item-three'
            }
        ]

        const element = createElement('component-dropdown', { is: Dropdown })
        element.options = {
            multiple: false,
            items,
            selectedItems: []
        }

        document.body.appendChild(element)

        return Promise.resolve().then(() => {
            expect(element).toMatchSnapshot();
            const allItemElements = element.shadowRoot.querySelectorAll('.item');
            expect(allItemElements).toHaveLength(items.length);

            items.forEach((item, index) => {
                const itemElement = allItemElements[index];
                expect(itemElement.textContent).toBe(item.title);
                expect(parseInt(itemElement.dataset.index, 10)).toBe(index);
            })
        })
    })

    it('fires selection event when item is clicked', () => {
        const items = [
            {
                title: 'First Selected Item',
                id: 'item-one'
            },
            {
                title: 'Second Item',
                id: 'item-two'
            }, {
                title: 'Third Item',
                id: 'item-three'
            }
        ]

        const element = createElement('component-dropdown', { is: Dropdown })
        element.options = {
            multiple: false,
            items,
            selectedItems: []
        }
        const evtListenerMock = jest.fn()
        element.addEventListener('selection', evtListenerMock)

        document.body.appendChild(element)

        return Promise.resolve().then(() => {
            const itemElement = element.shadowRoot.querySelector('.item');
            itemElement.click();
            expect(evtListenerMock).toHaveBeenCalledTimes(1);
        })
    })
})