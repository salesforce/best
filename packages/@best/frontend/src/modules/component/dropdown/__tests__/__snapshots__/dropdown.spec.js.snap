// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component-dropdown display correct text and data-index with one selected item 1`] = `
<component-dropdown>
  #shadow-root(open)
    <div
      class="control closed"
    >
      <div
        class="selected"
      >
        <img
          src="/resources/arrow.svg"
        />
        <div
          class="selected-item"
          data-index="0"
        >
          First Selected Item
        </div>
      </div>
      <div
        class="items"
      >
        <div
          class="item"
          data-index="0"
        >
          First Selected Item
        </div>
        <div
          class="item"
          data-index="1"
        >
          Second Item
        </div>
        <div
          class="item"
          data-index="2"
        >
          Third Item
        </div>
      </div>
    </div>
</component-dropdown>
`;

exports[`component-dropdown displays items in correct order 1`] = `
<component-dropdown>
  #shadow-root(open)
    <div
      class="control closed"
    >
      <div
        class="items"
      >
        <div
          class="item"
          data-index="0"
        >
          First Selected Item
        </div>
        <div
          class="item"
          data-index="1"
        >
          Second Item
        </div>
        <div
          class="item"
          data-index="2"
        >
          Third Item
        </div>
      </div>
    </div>
</component-dropdown>
`;

exports[`component-dropdown displays no items when none are provided 1`] = `
<component-dropdown>
  #shadow-root(open)
    <div
      class="control closed"
    />
</component-dropdown>
`;

exports[`component-dropdown displays no selected items when none are provided 1`] = `
<component-dropdown>
  #shadow-root(open)
    <div
      class="control closed"
    >
      <div
        class="items"
      >
        <div
          class="item"
          data-index="0"
        >
          First Selected Item
        </div>
        <div
          class="item"
          data-index="1"
        >
          Second Item
        </div>
        <div
          class="item"
          data-index="2"
        >
          Third Item
        </div>
      </div>
    </div>
</component-dropdown>
`;
